pennylane/labs/tests/dla/test_structure_constants_dense.py

```diff
@@ -0,0 +1,104 @@
+
+import pennylane as qml
+from pennylane.labs.dla import structure_constants_dense
+from pennylane.pauli import PauliSentence, PauliWord
+
+## Construct some example DLAs
+# TFIM
+gens = [PauliSentence({PauliWord({i: "X", i + 1: "X"}): 1.0}) for i in range(2)]
+gens += [PauliSentence({PauliWord({i: "Z"}): 1.0}) for i in range(3)]
+Ising3 = qml.pauli.lie_closure(gens, pauli=True)
+
+# XXZ-type DLA, i.e. with true PauliSentences
+gens2 = [
+    PauliSentence(
+        {
+            PauliWord({i: "X", i + 1: "X"}): 1.0,
+            PauliWord({i: "Y", i + 1: "Y"}): 1.0,
+        }
+    )
+    for i in range(2)
+]
+gens2 += [PauliSentence({PauliWord({i: "Z"}): 1.0}) for i in range(3)]
+XXZ3 = qml.pauli.lie_closure(gens2, pauli=True)
+
+
+class TestAdjointRepr:
+    """Tests for structure_constants"""
+
+    def test_structure_constants_dim(self):
+        """Test the dimension of the adjoint repr"""
+        d = len(Ising3)
+        Ising3_dense = np.array([qml.matrix(op, wire_order=range(3)) for op in Ising3])
+        adjoint = structure_constants_dense(Ising3_dense)
+        assert adjoint.shape == (d, d, d)
+        assert adjoint.dtype == float
+
+    def test_structure_constants_with_is_orthonormal(self):
+        """Test that the structure constants with is_orthonormal=True/False match for
+        orthogonal inputs."""
+
+        Ising3_dense = np.array([qml.matrix(op, wire_order=range(3)) for op in Ising3]) / np.sqrt(8)
+        adjoint_true = structure_constants_dense(Ising3_dense, is_orthonormal=True)
+        adjoint_false = structure_constants_dense(Ising3_dense, is_orthonormal=False)
+        assert np.allclose(adjoint_true, adjoint_false)
+
+    @pytest.mark.parametrize("dla", [Ising3, XXZ3])
+    @pytest.mark.parametrize("use_orthonormal", [False, True])
+    def test_structure_constants_elements(self, dla, use_orthonormal):
+        r"""Test relation :math:`[i G_\alpha, i G_\beta] = \sum_{\gamma = 0}^{\mathfrak{d}-1} f^\gamma_{\alpha, \beta} iG_\gamma`."""
+
+        d = len(dla)
+        dla_dense = np.array([qml.matrix(op, wire_order=range(3)) for op in dla])
+        if use_orthonormal:
+            gram_inv = sp.linalg.sqrtm(
+                np.linalg.pinv(np.tensordot(dla_dense, dla_dense, axes=[[1, 2], [2, 1]]).real)
+            )
+            dla_dense = np.tensordot(gram_inv, dla_dense, axes=1)
+            dla = [(scale * op).pauli_rep for scale, op in zip(np.diag(gram_inv), dla)]
+        ad_rep = structure_constants_dense(dla_dense, is_orthonormal=use_orthonormal)
+        for i in range(d):
+            for j in range(d):
+
+                comm_res = 1j * dla[i].commutator(dla[j])
+                res = sum((c + 0j) * dla[gamma] for gamma, c in enumerate(ad_rep[:, i, j]))
+                res.simplify()
+                assert set(comm_res) == set(res)  # Assert equal keys
+                if len(comm_res) > 0:
+                    assert np.allclose(*zip(*[(comm_res[key], res[key]) for key in res.keys()]))
```

Comment:
I think the tests are nice. We should add one on a non-orthonormal set of operators, though, that checks exactly this defining property of the structure constants :)

Reply:
(Can do this just by applying some random coefficients to the existing DLAs.)

Reply:
Ah yeah, I meant a non-diagonal coefficients matrix :) 
Like
```python
coeffs = np.random.random((len(dla), len(dla)))
new_dla = [qml.sum(*(c * op for c, op in zip(_coeffs, dla))) for _coeffs in coeffs]
```
or so

Reply:
I added a tested orthonormalize function but somehow the test now fails for the cases where I orthonormalize the basis 🤔 I am confused

Reply:
I think the problem is the inconsistency between the trace inner product for matrices and pauli sentences that is assumed. I.e. the pauli sentence one is normalized whereas in the matrix case we dont. So I think to have consistent adjoint representations we should fix one convention and stick to that. Not sure which one though?

```
x = X(0) @ X(1)
x = x.pauli_rep
(x @ x).trace()
# 1.0
xm = qml.matrix(x, wire_order=range(2))
np.trace(xm @ xm)
# 4.0
```

I feel like the normalized one is neater because it should not depend on the (potentially redundant) matrix dimensions. E.g. in the example if you set the wire_order to range(3) the inner product is 8

Reply:
Fully agree that we should be using $\text{tr}[A^\dagger B] / d$